kernel/kheap.c

```diff
@@ -93,9 +92,17 @@ void *k_heap_aligned_alloc(struct k_heap *h, size_t align, size_t bytes,
 			 */
 		}
 
-		(void) z_pend_curr(&h->lock, key, &h->wait_q,
-				   K_TICKS(end - now));
+		(void) z_pend_curr(&h->lock, key, &h->wait_q, timeout);
 		key = k_spin_lock(&h->lock);
+
+		if (!K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+			int64_t remaining = end - sys_clock_tick_get();
```

Comment:
Not loving the performance regression here: this is going to call back into the timer driver to recompute the next timeout at the end of the loop, even though we may very well succeed at the next allocation and not need the timeout.  It's also doing that math while holding the heap lock, which is needless (it's not touching heap data here).

Maybe what you want to do is retain the original structure and just fix the `end` value?  Something like this before the pend call?

```c
end = timeout == K_FOREVER ? INT64_MAX : end;
```

Reply:
Or a hybrid: retain your optimization of "skip the sys_clock_tick_get() if K_FOREVER", but do it before the pend instead of after.

Reply:
OK.  I think that needs to be before the loop or the initial check will fail.

Reply:
It is handy though on a low memory embedded platform.

Reply:
Rather than
`end = timeout == K_FOREVER ? INT64_MAX : end;`
I could do:
`end &= INT64_MAX;`
Then there's no branch.  Do you have a preference?

Reply:
I guess a UB pedant would say that the latter is making assumptions about 2's complement representations, but I'm not one.  No preference here.

Reply:
Went with your original proposal.  I think it's a little more clear what is being done.